add test for pthread_kill

nilCux opened the pull request:
Close #1716 Done with several problems.
- Send signal 0 with pthread_kill will return -1
- Send signal to terminated threads behave wierd (Even crashed when the signal is SIGALRM)

anton-bondarev commented on src/tests/posix/pthread/Mybuild:
```diff
@@ -35,3 +35,9 @@ module pthread_cond_timedwait_test {
+	source "pthread_kill_test.c"
+	depends embox.compat.posix.proc.signal
+	depends embox.compat.posix.pthreads
+}
```
a new line please

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,121 @@
+#include <signal.h>
+#include <embox/test.h>
+
+
```
rm empty lines

nilCux replied:
Sorry for messing up the code style. I will modify it carefully.

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,121 @@
+EMBOX_TEST_SUITE("posix/pthread_kill");
+
+
+void sighand(int signo)
```
please look through code-style
```
static void sighand(int signo) {
}
```

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,121 @@
+
+void *thread_test(void *parm)
+{
+    int rc = sleep(10);
```
sleep 10 sec? Is not too long?

nilCux replied:
How about 5 or 3? How many second is better?

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,121 @@
+
+void sighand(int signo)
+{
+    return;
```
empty handler? May be add some functionality in it?

nilCux replied:
May I add test_assert_not_zero(signo) here? Do you have any other suggestion? 
Thank you so much!

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,121 @@
+
+/*
+* Currently, using pthread_kill to send signal to terminated threads
+* behaves wierd. If the signal sent is SIGALRM, the program will crash.
```
Can you investigate it with more details?

nilCux replied:
I will dig deeper and send you a mail when I finish.

anton-bondarev replied:
Ok

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,109 @@
+EMBOX_TEST_SUITE("posix/pthread_kill");
+
+void sighand(int signo) {
+    test_assert_not_zero(signo);
```
I mean than you can check that right signal handler has been called. Using different handlers for example or set some variables

nilCux replied:
I emit different characters in two different handlers. I set them as handler for different Signals. I send only one Signal and capture it after join() with test_assert_get_emitted(). So I can see if the handler called is good.

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,109 @@
+    return;
+}
+
+void *thread_test(void *parm) {
```
please make internal routines static

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -0,0 +1,109 @@
+    sigemptyset(&actions.sa_mask);
+    actions.sa_flags = 0;
+    actions.sa_handler = sighand;
+    rc = sigaction(SIGALRM,&actions,NULL);
```
Please use spases
(SIGALRM,&actions,NULL)->(SIGALRM, &actions, NULL)

nilCux replied:
Thank you so much for your patience. I will be more careful.

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -15,36 +15,49 @@
 
-void *thread_test(void *parm) {
+static void sighand_b(int signo) {
+    test_assert_not_zero(signo);
```
test emit is enough in my mind

anton-bondarev commented on src/tests/posix/pthread/pthread_kill_test.c:
```diff
@@ -15,36 +15,49 @@
 
-void *thread_immediate_return(void *parm) {
+/*
+static void *thread_immediate_return(void *parm) {
```
move it please where it's used

nilCux replied:
I wanted to use it to generate a thread that terminates immediately. Since it returns immediately, we can guarantee the thread is terminated. Then we can test sending signal to a terminated thread.

Do you think I should delete it or leave it for later use?

anton-bondarev replied:
I'm not sure. But we try to add only used code.
As for this situation, maybe better way is to create a new issue and described the problem in it. Delete unused code to clean the PR

nilCux replied:
How about I delete unused code, and mention the problem in another issue when I finish investigating the problem about sending signal to terminated threads?

anton-bondarev commented:
@nilCux Thanks
Can you also add the test to x86/test/unit template?

anton-bondarev commented:
@nilCux Now your tests are worked?

nilCux commented:
> @nilCux Now your tests are worked?

When I run this test, it will fail because the return value for sending signal 0 is -1.
I asked in #1716 , based on alexkalmuk 's reply, I kept the test that way.

```
test: running embox.test.posix.pthread.pthread_kill_test .
        failure at src/tests/posix/pthread/pthread_kill_test.c : 121, in function __test_case_at_line_107
                test_assert_zero(rc)
           case at src/tests/posix/pthread/pthread_kill_test.c : 107
                "Send signal 0 to a thread"

        testing pthread_kill_test (posix/pthread_kill) failed
                1/2 failures

```

nilCux commented:
To make the unit test work, I also fixed #1808 . I sent you a mail with more details.

For signal.c , I respected the original coding style with blank lines between blocks.
Also, I added back the part we deleted last time (It works now) and added another test case for invalid Signal.

If I made any mistake, thank you for letting me know. Btw, speaking of is functionality it works on my pc with all tests passed.

anton-bondarev commented:
@nilCux Thanks a lot.

The only one things obstruct me to approve, Can you please extract commits with signal.c modification to a new PR?

nilCux commented:
Hi, sorry about the mess. I tried to squash my commits with git rebase. Evidently, I messed up.
For my code, I only changedan letter from 'a' to 'b'. It was an typing error.

nilCux commented:
What I did was:
```
git checkout my-branch
git fetch upstream
git rebase upstream/master
```
with
```
# upstream  https://github.com/embox/embox.git (fetch)
# upstream  https://github.com/embox/embox.git (push)
```
I find there are 100+ files changed after my operation. I'm glad to re-open another PR and close this one if it could help.

anton-bondarev commented:
@nilCux 
> After that, when I try to push, git asked me to pull. So I pilled and pushed.
You have not to pull only force push after rebase
```
git fetch upstream
git rebase upstream/master
git push origin +1716-unit-test-pthread_kill

```

Deryugin commented:
@nilCux that looks strange. Can you try to synchronize your master with upstream master?

```
git checkout master
git pull upstream master
```

That should produce no merge commits. Then
```
git checkout 1716-unit-test-pthread_kill
git rebase master
```

Hope that helps.

nilCux commented:
Thank you so much for your help. @Deryugin @anton-bondarev 
It's ok now.

nilCux commented:
> @nilCux Thanks a lot.
> 
> The only one things obstruct me to approve, Can you please extract commits with signal.c modification to a new PR?

I opened #1838 for the change in signal.c and pthread.h (pthread_kill was declared here).
However, I left those changes in this PR too, because without those change, there will be a crash during our tests.

nilCux commented:
Hi, since the CI build didn't pass, I ran it again in my own repository. For the first time, it failed at x86/test/net , too. But when I re-run this single build, it passed. I wonder if I made any mistake?

alexkalmuk commented:
> Hi, since the CI build didn't pass, I ran it again in my own repository. For the first time, it failed at x86/test/net , too. But when I re-run this single build, it passed. I wonder if I made any mistake?

The `x86/test/net` (I guess ping test) is not very stable now somewhy, so it's not your mistake.

anton-bondarev commented:
@nilCux Hi
Remove changes from signal.c please (add a new commit with reverting changes) and then squash

I investigate the problems of zero signal and stopped thread and it's more difficult than your suggestion. I offer to merge your tests (they are good) and solve the problems in your PR #1838

anton-bondarev commented:
@nilCux Also it's great if there are three commits in your PR, add phread_kiil to pthread.h, add unit tests, and add your test to x86/test/unit template

nilCux commented:
Just did as you asked, please verify if thats what I should do.
Thank you again for all your help during these days. Take good care of your health since the covid-2019 is still spreading. I will definitely try my best in the official application process that starts tomorrow for the project about memory profiler, so, I hope I can contribute more for Embox. Bests.

anton-bondarev commented:
@nilCux Thank you
I've merged it to master.

> Take good care of your health
Thanks, and you!
